docs/source/design.rst

```diff
@@ -1,7 +1,201 @@
+
+The ``input_type`` member is a list of rules that defines the types of
+the inputs to the operation. This is sometimes called the signature.
+
+The ``output_type`` member is a rule or a method that defines the output type
+of the operation. This is sometimes called the return type.
+
+An example of ``input_type``/``output_type`` usage is the
+:class:`~ibis.expr.operations.Log` class:
+
+.. code-block:: python
+
+   class Log(Node):
+
+       input_type = [
+           rules.double(),
+           rules.double(name='base', optional=True)
+       ]
+       output_type = rules.shape_like_arg(0, 'double')
+
+This class describes an operation called ``Log`` that takes one required
+argument: a double scalar or column, and one optional argument: a double scalar
+or column named ``base`` that defaults to nothing if not provided. The base
+argument is ``None`` by default so that the expression will behave as the
+underlying database does.
+
+These objects are instantiated when you use ibis APIs:
+
+.. code-block:: python
+
+   >>> import ibis
+   >>> t = ibis.table([('a', 'double')])
+   >>> log_1p = (1 + t.a).log()  # an Add and a Log are instantiated here
+
+.. _expr_vs_ops:
+
+Expressions vs Operations: Why are they different?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Separating expressions from their underlying operations makes it easy to
+generically describe and validate the inputs to particular nodes. In the log
+example, it doesn't matter what *operation* (node) the double-valued arguments
+are coming from, they must only satisfy the requirement denoted by the rule.
+
+Separation of the :class:`~ibis.expr.types.Node` and
+:class:`~ibis.expr.types.Expr` classes also allows the API to be tied to the
+physical type of the expression rather than the particular operation, making it
+easy to define the API in terms of types rather than specific operations.
+
+.. _compilation:
+
+Compilation
+-----------
+
+The next major component of ibis is the compilers.
+
+The first few versions of ibis directly generated strings, but the compiler
+infrastructure was generalized to support compilation of `SQLAlchemy
+<https://docs.sqlalchemy.org/en/latest/core/tutorial.html>`_ based expressions.
+
+The compiler works by translating the different pieces of SQL expression into a
+string or SQLAlchemy expression.
+
+The main pieces of a ``SELECT`` statement are:
+
+#. The set of column expressions (``select_set``)
+#. ``WHERE`` clauses (``where``)
+#. ``GROUP BY`` clauses (``group_by``)
+#. ``HAVING`` clauses (``having``)
+#. ``LIMIT`` clauses (``limit``)
+#. ``ORDER BY`` clauses (``order_by``)
+#. ``DISTINCT`` clauses (``distinct``)
+
+Each of these pieces is translated into a SQL string and finally assembled by
+the instance of the :class:`~ibis.sql.compiler.ExprTranslator` subclass
+specific to the backend being compiled. For example, the
+:class:`~ibis.impala.compiler.ImpalaExprTranslator` is one of the subclasses
+that will perform this translation.
```

Comment:
May be worth noting explicitly that SQL is only one target, however the library was designed with first-class SQL support in mind, being the *lingua franca* of analytics.